Support schemas with postconditions on query parameters

quimrstorres opened the pull request:
In some cases, the schema records include a few more keys besides the model
definition itself. For example, schemas with postconditions include the
keys `:postconditions` and `:post-name`. In cases like that, the keys
need to be excluded before the conversion to Swagger spec.

Since the keys unrelated to the model definition were not being ignored,
the use of schemas with postconditions in the query parameters was causing
an exception to be raised.
The fields are now excluded by using `schema-tools.core/schema-value`
before the conversion to Swagger spec.

Closes #104

Deraen commented on src/ring/swagger/swagger2.clj:
```diff
@@ -51,7 +52,7 @@
 (defmethod extract-parameter :default [in model options]
   (if model
-    (for [[k v] (-> model common/value-of rsc/strict-schema)
+    (for [[k v] (-> model stc/schema-value common/value-of rsc/strict-schema)
```
Hmm, schema-value and value-of try to do similar things.

I wonder if using schema-value here loses some information?

quimrstorres replied:
In the case depicted in the issue, the return from `value-of` is the model passed as parameter, which is different from what is returned from `schema-value`. The goal of using it here is in fact excluding some information from the model that is not relevant for the conversion (and that was causing the exception).

I am not sure that it never excludes relevant information though. Do you have any suspicion in particular that I can investigate? Or any suggestion to make sure that doesn't happen?

Deraen replied:
Try `:query (s/maybe {:whatever s/Str})`.

Maybe schema should lead to `:allowEmptyValue true` in JSON-Schema, but as `schema-value` just walks over `maybe` this is lost?

Deraen replied:
@ikitommi Any ideas?

I think `extract-parameter` should directly call `->swagger` (like it does for `:body` schemas) so the top-most Schema objects are correctly analyzed. Or maybe maybe there is something here I'm missing? I'm not very familar on what Swagger Spec says about `:query etc schemas.

ikitommi replied:
Hmm... the `:query (s/maybe {:whatever s/Str})` is a extra special case, and it should/could cause all keys to be optional within? Anyway, we don't seem to be handling that at the moment, the more normal maybe-case should work with the patch too: `:query {:whatever (s/maybe s/Str)})`.

@Deraen - non-body-parameters can't have `:schema`s, the schema keys need to be flat/primitive types. I think the suggestion works, but `stc/schema-value` & `common/value-of` should be in different order. For some reason, we are supporting vars too, so one could say `:query #'Adult` and that would break with this code.

ikitommi replied:
Actually, the `:query (s/maybe {:whatever s/Str})` fails too, for the same reason as Constrained, this will most likely fix that too.

(ss2/swagger-json {:paths {"/" {:get {:parameters {:query (s/maybe {:a s/Str})}}}}})
; java.lang.IllegalArgumentException: don't know how to convert {:a java.lang.String} into a Swagger Schema. Check out ring-swagger docs for details.

quimrstorres replied:
Just tried it out and the patch seems to solve it. The resulting Swagger spec looks like this:

```clj
(rs/swagger-json {:paths {"/" {:get {:parameters {:query (schema/maybe {:a schema/Str})}}}}})
; {:swagger "2.0"
;  :info {:title "Swagger API", :version "0.0.1"}
;  :produces ["application/json"]
;  :consumes ["application/json"]
;  :paths {"/" {:get {:parameters [{:in "query"
;                                   :name "a", :description ""
;                                   :required true
;                                   :type "string"}]
;                     :responses {:default {:description ""}}}}}, :definitions {}}
```

It does not include the `:allowEmptyValue` but I guess that information is inherently lost with the separation into primitive types.

The other, more common, example mentioned by @ikitommi already includes the `:allowEmptyValue`:

```clj
(rs/swagger-json {:paths {"/" {:get {:parameters {:query {:a (schema/maybe schema/Str)}}}}}})
; {:swagger "2.0"
;  :info {:title "Swagger API", :version "0.0.1"}
;  :produces ["application/json"]
;  :consumes ["application/json"]
;  :paths {"/" {:get {:parameters [{:in "query"
;                                   :name "a"
;                                   :description ""
;                                   :required true
;                                   :type "string"
;                                   :allowEmptyValue true}]
;                     :responses {:default {:description ""}}}}}, :definitions {}}
```

Since these examples came up in the discussion, would it make sense to add them in tests?

ikitommi replied:
Looks great! And tests are good, to avoid future regression.

quimrstorres commented:
I switched the order of `value-of` and `schema-value`, added the tests and changed the commit message to be accurate according to these new findings. Can you take a look again?

ikitommi commented:
Looks good, thanks!!

ikitommi commented:
Updated README & pushed out new `[metosin/ring-swagger "0.23.0-SNAPSHOT"]`. The immutable version for this is `[metosin/ring-swagger "0.23.0-20160817.111844-3"] if you need it in non-prod. We'll add some more stuff as next will be a MINOR update.

quimrstorres commented:
Great! Thanks for the quick update. 👍
